src/d3d9/shaders/d3d9_fixed_function_frag.frag

```diff
@@ -0,0 +1,667 @@
+layout(set = 0, binding = 13) uniform texture3D t3d[TextureStageCount];
+
+layout(origin_upper_left) in vec4 gl_FragCoord;
+
+layout(set = 15, binding = 0) uniform sampler sampler_heap[];
+
+
+// Functions to extract information from the packed texture stages
+uint colorOp(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[0], 0, 5);
+}
+uint colorArg0(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[0], 5, 6);
+}
+uint colorArg1(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[0], 11, 6);
+}
+uint colorArg2(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[0], 17, 6);
+}
+
+uint alphaOp(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[0], 23, 5);
+}
+uint alphaArg0(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[1], 0, 6);
+}
+uint alphaArg1(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[1], 6, 6);
+}
+uint alphaArg2(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[1], 12, 6);
+}
+
+bool resultIsTemp(uint stageIndex) {
+    return bitfieldExtract(data.Stages[stageIndex].Primitive[1], 18, 1) != 0;
+}
+bool globalSpecularEnabled() {
+    return bitfieldExtract(data.Stages[0].Primitive[1], 19, 1) != 0;
+}
+
+
+vec4 calculateFog(vec4 vPos, vec4 oColor) {
+    vec3 fogColor = vec3(rs.fogColor[0], rs.fogColor[1], rs.fogColor[2]);
+    float fogScale = rs.fogScale;
+    float fogEnd = rs.fogEnd;
+    float fogDensity = rs.fogDensity;
+    D3DFOGMODE fogMode = specUint(SpecPixelFogMode);
+    bool fogEnabled = specBool(SpecFogEnabled);
+    if (!fogEnabled) {
+        return oColor;
+    }
+
+    float w = vPos.w;
+    float z = vPos.z;
+    float depth = z * (1.0 / w);
+    float fogFactor;
+    switch (fogMode) {
+        case D3DFOG_NONE:
+            fogFactor = in_Fog;
+            break;
+
+        // (end - d) / (end - start)
+        case D3DFOG_LINEAR:
+            fogFactor = fogEnd - depth;
+            fogFactor = fogFactor * fogScale;
+            fogFactor = spvNClamp(fogFactor, 0.0, 1.0);
+            break;
+
+        // 1 / (e^[d * density])^2
+        case D3DFOG_EXP2:
+        // 1 / (e^[d * density])
+        case D3DFOG_EXP:
+            fogFactor = depth * fogDensity;
+
+            if (fogMode == D3DFOG_EXP2)
+                fogFactor *= fogFactor;
+
+            // Provides the rcp.
+            fogFactor = -fogFactor;
+            fogFactor = exp(fogFactor);
+            break;
+    }
+
+    vec4 color = oColor;
+    vec3 color3 = color.rgb;
+    vec3 fogFact3 = vec3(fogFactor);
+    vec3 lerpedFrog = mix(fogColor, color3, fogFact3);
+    return vec4(lerpedFrog.r, lerpedFrog.g, lerpedFrog.b, color.a);
+}
+
+
+// [D3D8] Scale Dref to [0..(2^N - 1)] for D24S8 and D16 if Dref scaling is enabled
+vec4 adjustDref(vec4 texCoord, uint referenceComponentIndex, uint samplerIndex) {
```

Comment:
readability / potential code gen meme nit, i'd prefer if this just consumed and returned a `float` so usage code just does something like `a.z = adjustDref(coord.z, ...)`